src/gax/src/http_client/mod.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::error::Error;
+use crate::error::HttpError;
+use auth::Credential;
+type Result<T> = std::result::Result<T, crate::error::Error>;
+
+#[derive(Clone)]
+pub struct ReqwestClient {
+    inner: reqwest::Client,
+    cred: Credential,
+    endpoint: String,
+}
+
+impl ReqwestClient {
+    pub async fn new(config: ClientConfig, default_endpoint: &str) -> Result<Self> {
+        let inner = reqwest::Client::new();
+        let cred = if let Some(c) = config.cred {
+            c
+        } else {
+            ClientConfig::default_credential().await?
+        };
+        let endpoint = config
+            .endpoint
+            .unwrap_or_else(|| default_endpoint.to_string());
+        Ok(Self {
+            inner,
+            cred,
+            endpoint,
+        })
+    }
+
+    pub fn builder(&self, method: reqwest::Method, path: String) -> reqwest::RequestBuilder {
+        self.inner
+            .request(method, format!("{}{path}", &self.endpoint))
+    }
+
+    pub async fn execute<I: serde::ser::Serialize, O: serde::de::DeserializeOwned>(
+        &self,
```

Comment:
@codyoss this version of `execute()` may have wrecked your plans for pagination.  If you can point me to some WIP with how you were planning to use it the version that consumed a `token: String`, I can make sure to adapt my current code.

Or even some pseudo-code about why consuming a `self` was not a great idea.

Reply:
The issue comes when putting any code within a higher order function in rust. When you do this you need to move variables to new code stacks. I was having issues making the compiler happy if self was ever mentioned in the function for this reason -- which does make sense I suppose. If you think about it the list RPCs will return a type out-side the scope of the call to the client, so the execute function passed to the paginator could live longer than the client itself. So it all has to do with lifetimes. You could replicate this with any lambda I assume, or a func that takes one of the Fn traits.

```
    // build up a request
        gax::paginator::Paginator::new(tok, move |token| {
            // do some non-async things
            async move {
                // call the execute function that refs self and see a compile error.
                self.execute(...)
            }
        })
```

Reply:
It could be I was holding it wrong here too. Would be happy to get some more input if I am simply missing something. All the moving done with the higher order functions was tripping me up a bit.

Reply:
PTAL.  In generated code there is an extra layer of indirection (the `client::Foo` and `traits::dyntraits::Foo`), but I think that will make things faster (clones are cheaper), and not affect the lifetime analysis.